### DISCUSSION TITLE
auto optimization

### DISCUSSION THREAD

**Discussion by HilkopterBob** [zauberzeug/nicegui#682] [Q&A]
hey is there a possibility within nicegui to start an automated optimization job?

**Comment by falkoschindler** [comment]
What do you mean with "automated optimization job"?

**Reply by HilkopterBob** [reply]
To better describe my situation:

The nicegui program I wrote is a bit sluggish from time to time. This is mainly due to the fact that animations such as "hover" sometimes lag behind by fractions of a second in the checkboxes. And that with little to no operations running in the background.

That's why I'm wondering whether there is a command that builds the nicegui application similar to a compiler and then also optimizes it in the direction of faster loading times, eliminating unnecessary/avoidable operations, etc.

**Reply by rodja** [reply]
Most of the time sluggishness is caused by a slow internet connection. The whole backend-first approach from NiceGUI relies on a fast internet connection and therefore is not meant to replace classical web apps.
Another regular cause could be too much work on the main thread. With NiceGUI you really need to move all IO/CPU bound delays behind async/await. Otherwise the server can not do anything else. And because every user-interaction is handled there, it will feel like you are back in the internet of the 90ties.

**Reply by HilkopterBob** [reply]
thanks for the explanation @rodja, i will try implementing the async behavior into my app and write back if it helped.

besides that, i test my project on my mashine, so the connection speed isn't a problem afaik. Do you have more tips?

**Reply by rodja** [reply]
@HilkopterBob You could create a stripped down version of your code? Eg. try, if a simple checkbox printing something to the console works. Then you could step by step reintroduce more of your code. This way you can discover where the problems are being introduced.
